dnfdaemon-server/session_manager.cpp

```diff
@@ -93,6 +93,10 @@ void SessionManager::on_name_owner_changed(sdbus::Signal & signal) {
 }
 
 sdbus::MethodReply SessionManager::open_session(sdbus::MethodCall & call) {
+    if (!active) {
+        throw sdbus::Error(dnfdaemon::ERROR, "Cannot open new session.");
+    }
+
```

Comment:
So just one more thing on the concurrency synchronization. There's a race that can occur if a termination signal is received right at this point:
- The check above completed without throwing, `active` was `true`.
- Termination signal arrived at this point, the sessions were cleared
- Execution continues from this point and a new session is created

I'm not sure if you want to use the same mutex that you use for the sessions as I don't know exactly what's the concurrency situation there, but it appears so, as clearing the sessions seems to be the main objective of `shut_down()`.

Alternatively you could have an extra mutex for the sychronization of the `active` state.

On a side note if there wasn't the need for a mutex, it would still be formally correct to have the `active` flag wrapped in `std::atomic`, since you are reading/modifying it from two separate execution flows (the main one and then a signal handler). Granted a boolean can only have two values, so I'm thinking no matter the point at which you interrupt the writing of a new value, the reader will always read `true` or `false`, not anything halfway through... but anyway :)